api/api.go

```diff
@@ -1469,19 +1473,75 @@ func (api *Server) getLogsInBlock(filter *LogFilter, start, count uint64) ([]*io
 		end = api.bc.TipHeight()
 	}
 	for i := start; i <= end; i++ {
-		header, err := api.dao.HeaderByHeight(i)
+		logsInBlock, err := api.getLogsInBlock(filter, i)
 		if err != nil {
-			return logs, status.Error(codes.InvalidArgument, err.Error())
+			return logs, err
 		}
-		if !filter.ExistInBloomFilter(header.LogsBloomfilter()) {
-			continue
+		if logsInBlock != nil {
+			logs = append(logs, logsInBlock...)
 		}
-		receipts, err := api.dao.GetReceipts(i)
+	}
+	return logs, nil
+}
+
+func (api *Server) getLogsInBlock(filter *l.LogFilter, blockNumber uint64) ([]*iotextypes.Log, error) {
+	header, err := api.dao.HeaderByHeight(blockNumber)
+	if err != nil {
+		return nil, status.Error(codes.InvalidArgument, err.Error())
+	}
+	if !filter.ExistInBloomFilter(header.LogsBloomfilter()) {
+		return nil, nil
+	}
+	receipts, err := api.dao.GetReceipts(blockNumber)
+	if err != nil {
+		return nil, status.Error(codes.InvalidArgument, err.Error())
+	}
+	return filter.MatchLogs(receipts), nil
+}
+
+// TODO: improve using goroutine
+func (api *Server) getLogsInRange(filter *l.LogFilter, start, end uint64) ([]*iotextypes.Log, error) {
+	if start > 0 {
+		start = 0 // from the beginning
+	}
```

Comment:
i am confused here. say a contract have more than `RangeQueryLimit` number of logs, so we want to query range by range. but with this check which force the start to be 0, we cannot query range by range then

Reply:
I addressed your comment on the latest commit. Now RangeQuery does not force the start height to be 0.